### PR TITLE
qtwayland: declare proper `app_id` for wrapped executables

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

Since NixOS 21.11 it seems as if QT uses Wayland if possible[1].
However, my `pinentry-qt` flavor stopped floating because it's now
running in Wayland-mode rather than in XWayland mode where this seems to
be fine.

I wanted to add a rule to my `sway(1)`-config for that, but realized
that `pinentry` is missing an `app_id` to match:

    $ swaymsg -t get_tree | jq '.nodes[2].nodes[1].nodes[1].nodes[1].app_id'
    ""

This is because `QWaylandWindow::initWindow()` uses the application's
`baseName` to determine the app id. Unfortunately the `baseName`
drops all chars of the filename after the first dot[2]. This means that
every wrapped Nix package (i.e. `pkgs.foo` with `$out/bin/.foo-wrapped`)
will have an empty-string as baseName because the first character of the
filename is a dot. Since we're using the `wrapQtAppsHook` quite
excessively, a lot of programs are affected by this.

In order to work around this, I implemented a small patch for
`qtwayland` that strips away the `nixpkgs`-specific `.(name)-wrapped` of
a filename if needed and then sets the `app_id` to the expected
`baseName`. This is useful to make e.g. `sway`-configs with
`for_window`[3]-expressions from other distros compatible.

With this change, the `app_id` is set as I'd expect it:

    $ swaymsg -t get_tree | jq '.nodes[2].nodes[1].nodes[1].nodes[1].app_id'
    "pinentry-qt"

Even though we'll need the patch to get e.g. `foo` from `.foo-wrapped`,
I decided to file a bug-report against upstream[4].

[1] https://nixos.org/manual/nixos/stable/release-notes.html#sec-release-21.11
[2] https://doc.qt.io/qt-5/qfileinfo.html#baseName
[3] https://man.archlinux.org/man/sway.5.en
[4] https://bugreports.qt.io/browse/QTBUG-99137

__Note:__ I really don't want to decide this on my own, but since it's technically a bugfix, I'd like to raise the question whether it's suitable for a backport (in case it's sufficiently reviewed & tested first).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by lheckemann — APPROVED** [review]
The comment could be in the patch instead (make it a commit and the comment the commit message?). Otherwise LGTM :+1: 

EDIT: also staging?

**Comment by Ma27** [comment]
@lheckemann moved the comment to the patch, based onto staging :)

**Comment by samueldr** [comment]
We probably want to consider adding a "libQtNixOS" thing, even if it's a header-only library, where we can implement these kind of things once.

This is the implementation in kwin:

 - https://github.com/NixOS/nixpkgs/blob/17336f60ffafb332cf27c9d8dd6e0aed54a95167/pkgs/desktops/plasma-5/kwin/0001-NixOS-Unwrap-executable-name-for-.desktop-search.patch#L53-L59

I figure it woule be best if we can at least sync-up the implementation. This one has been reviewed by @ttuegel.

**Note**: I have not comparatively "reviewed" your implementation. I'm not a C/C++/Qt expert, so I don't know which one would be "best". I'm only bringing up my concern over multiple implementations of the same concept starting to spring up left and right in the Qt/Plasma ecosystem

**Comment by Ma27** [comment]
So your suggestion is to effectively use `unwrapExecutablePath` rather than inlining the implementation as it's done in the current patch?

Seems to be a fine solution for me, should we create a repo for this library in the `NixOS/`-org, then? Also, would be interested in @ttuegel's opinion on this :)

**Comment by samueldr** [comment]
My suggestion was more about *starting to think about that*, but yes.

Using a unified implementation would be safer against "drift" if for some reason there was any difference in the actual behaviour of the different implementations.

I don't really know what the proper way forward would be, though.

* * *

By the way, assuming this current change fixes things, I don't mind it being merged, and ear-marked for future unification. Don't block improving support for Qt things under wayland on ideal implementation!

**Comment by Ma27** [comment]
> My suggestion was more about starting to think about that, but yes.

Fair. As soon as this is merged, I'd open an issue to keep track of this.

While the patch works fine for me, I'm only using a hand full of QT applications (most notably pinentry-qt where I *really* want a floating in sway), but I'm not sure if this could cause unexpected issues for e.g. KDE/Plasma users (cc @NixOS/qt-kde for a review).

**Comment by alyssais** [comment]
Just FWIW: Wayland-native pinentry-qt used to float in Sway just fine.  It's only stopped working in the last few months or so.

**Comment by lheckemann** [comment]
@alyssais are you sure? Mine has only ever floated correctly when using Xwayland.

**Comment by alyssais** [comment]
Yeah, I'm sure.  In fact, it still does, except for the window that asks me my smartcard passphrase (the window asking me to insert the smartcard floats, and the passphrase one used to but doesn't since [`85b180f1b014`](https://dev.gnupg.org/rP85b180f1b0148a07d51d244deb1a1245c9008cc3) ("qt: Show hint if passphrase is shown and formatting is enabled")).  I tested this config in a stateless VM:

```nix
{ pkgs, ... }:

let
  gpg-agent-conf = pkgs.writeText "gpg-agent.conf" ''
    pinentry-program ${pkgs.pinentry.qt}/bin/pinentry
  '';
in

{
  systemd.tmpfiles.rules = [
    "d /home/qyliss/.gnupg 0700 qyliss users"
    "L+ /home/qyliss/.gnupg/gpg-agent.conf - - - - ${gpg-agent-conf}"
  ];

  environment.etc."sway/config.d/qyliss.conf".source = pkgs.writeText "qyliss.conf" ''
    input * xkb_layout dvorak
    exec foot
  '';

  environment.systemPackages = with pkgs; [ gnupg ];

  programs.sway.enable = true;
  programs.sway.extraPackages = with pkgs; [ foot ];
}
```

**Comment by Ma27** [comment]
@alyssais I guess I can confirm your observations (passphrase forms don't float, but e.g. the "passphrase isn't safe, select a new one?"-window does - yes I gave the gpg key I generated in a VM for testing purposes a weak passphrase :sweat_smile: ).

And while it's actually quite nice that you seem to have fixed the issue for pinentry-qt (ref https://dev.gnupg.org/D549 for everyone following the thread), I still think there's still an issue with qtwayland on NixOS: I'd like to be able to configure any kind of window in my sway with some sort of matching mechanism, e.g. the `app_id` and this doesn't work for QT applications very well because $0 is called `.foo-wrapped` when using `wrapQTAppsHook`, so I still think that this patch is relevant.

ping @ttuegel any chance to get a review from you? I'm actually using this patch since December on my laptop and it works fine for me :)

**Review by lheckemann — APPROVED** [review]

**Pull request merged**
